usl_models/usl_models/atmo_ml/datasets.py

```diff
@@ -227,17 +234,112 @@ def data_generator():
-        dataset = dataset.shuffle(buffer_size=len(spatiotemporal_file_names))
-
+        dataset = dataset.shuffle(buffer_size=1000)  # Use a fixed buffer size
     dataset = dataset.batch(batch_size)
 
+    # Use fixed sizes for splitting the dataset
+    total_size = 1000  # Adjust this value based on your actual dataset size
+    train_size = int(0.7 * total_size)
+    val_size = int(0.15 * total_size)
+    test_size = total_size - train_size - val_size
+
+    train_dataset = dataset.take(train_size)
+    val_dataset = dataset.skip(train_size).take(val_size)
+    test_dataset = dataset.skip(train_size + val_size).take(test_size)
+
+    return train_dataset, val_dataset, test_dataset
+
+
+def load_prediction(
+    data_bucket_name: str,
+    spatiotemporal_folder: str,
+    spatial_folder: str,
+    lu_index_folder: str,
+    time_steps_per_day: int,
+    batch_size: int = 4,
+    storage_client: storage.Client = None,
+    firestore_client: firestore.Client = None,
+) -> tf.data.Dataset:
+    def pad_or_truncate_data(data, target_length, pad_value=0):
+        """Pad or truncate data to the target length along the first axis."""
+        current_length = data.shape[0]
+        if current_length > target_length:
+            return data[:target_length]
+        elif current_length < target_length:
+            pad_shape = [target_length - current_length] + list(data.shape[1:])
+            padding = np.full(pad_shape, pad_value, dtype=data.dtype)
+            return np.concatenate([data, padding], axis=0)
+        return data
+
+    # Load spatial, LU index, and spatiotemporal data from their respective folders
+    lu_index_data = load_lu_index_from_cloud(
+        data_bucket_name, lu_index_folder, storage_client, firestore_client
+    )
+    spatial_data = load_spatial_data_from_cloud(
+        data_bucket_name, spatial_folder, storage_client, firestore_client
+    )
+    spatiotemporal_data_list = load_spatiotemporal_data_from_cloud(
+        data_bucket_name, spatiotemporal_folder, storage_client, firestore_client
+    )
+
+    def data_generator():
+        # Iterate through each spatiotemporal data file
+        for spatiotemporal_data in spatiotemporal_data_list:
+            # Divide data into days and apply padding or truncation
+            # Divide the spatiotemporal data into daily inputs
+            inputs, _ = cnn_inputs_outputs.divide_into_days(
+                spatiotemporal_data, labels=None
+            )
+            for day_inputs in inputs:
+                day_inputs_padded = pad_or_truncate_data(
+                    day_inputs.numpy(), time_steps_per_day
+                )
+                yield {
+                    "spatiotemporal": day_inputs_padded,
+                    "spatial": spatial_data.numpy(),
+                    "lu_index": lu_index_data.numpy(),
+                }
+
+    # Create the dataset using the data generator
+    dataset = tf.data.Dataset.from_generator(
+        data_generator,

```

Comment:
For my understanding, why do we specify an output signature in this dataset but not in the above datasets?